Refactor AdSense setup components

felixarntz opened the pull request:
## Summary

Addresses issue #1014

## Relevant technical choices

* Instead of relying on passing AdSense frontend URLs from PHP to JS (some of these are slightly incorrect anyway), utility functions are added in JS to generate these URLs. These utility functions also support passing the Google account email address in the URL, which wasn't happening before, resulting in a more seamless user experience for people with multiple Google accounts.
* https://github.com/google/site-kit-wp/pull/1410/commits/0b969e66099bab815f44f00a3f38721f1ab84346 in particular fixes a problem with the current implementation which can have a user end up in pending state without having fetched a client ID, causing the snippet for such users not to be rendered (as a reminder, the AdSense snippet must be rendered in the frontend as soon as the user is pending or graylisted).

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

felixarntz commented on assets/js/components/settings/settings-module.js:
```diff
@@ -340,7 +336,6 @@ class SettingsModule extends Component {
 									{
 										hasSettings && ! setupComplete &&
 											<ModuleSetupIncomplete
-												screenID={ screenID }
```
The only reason this was still being passed is because the old `AdSenseSettingsStatus` component used it. Now that that is gone, we can eliminate this prop.

felixarntz commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+	useEffect( () => {
+		// Only submit changes if valid.
+		// TODO: Remove temporary hack to avoid saving in Storybook.
+		if ( ! canSubmitChanges || global.__STORYBOOK_ADDONS ) {
```
@aaemnnosttv @tofumatt I wanted to avoid firing requests like this in storybook and added this hack. 😆 Do you have any better idea?

aaemnnosttv replied:
We have some kind of fetch mock installed which is usable in Storybook, although we're not currently using it I don't think so you might need to look up how to use it. I forget exactly what it's called, it's named very similar to the one we use in Jest but slightly different 😆 - @tofumatt can you point him to the relevant docs?

tofumatt replied:
Mocking fetch _in_ storybook seems the right approach—I think @aaemnnosttv did that already in the Analytics code. But I can't find it now that I'm looking for it. If @aaemnnosttv knows off the top of his head that's cool, otherwise I'll circle back to this later 😄

tofumatt replied:
The more generic `fetch-mock` library we have loaded in storybook is: https://www.npmjs.com/package/fetch-mock. API docs are here: http://www.wheresrhys.co.uk/fetch-mock/ and there's a quick start here: https://github.com/wheresrhys/fetch-mock/blob/master/docs/cheatsheet.md

Looks like the very basics are:

```js
fetchMock
  .get('/expected-url', {status: 200, statusText: 'responseText' });
```

http://www.wheresrhys.co.uk/fetch-mock/#api-mockingget_post goes more in-depth. That should allow you to remove all the `global.__STORYBOOK_ADDONS` checks though. If it's not behaving well happy to pair with you or check out the code that isn't working if you want to push a WIP commit 🙂

aaemnnosttv commented on assets/js/googlesitekit/data/create-settings-store.js:
```diff
@@ -202,9 +202,16 @@ export const createSettingsStore = ( type, identifier, datapoint, {
-			invariant( values, 'values is required.' );
 			let response, error;
 
+			if ( ! values ) {
+				error = {
+					message: 'values is required.',
+				};
+				global.console.error( error.message );
+				return { response, error };
+			}
```
Shouldn't we check values in `actions.saveSettings` with invariant instead?

felixarntz replied:
I think it would be better to check for validity in `fetch*` since that's what actually needs the value. However, there was a problem with using `invariant` in there which I've already discussed a few weeks ago with @tofumatt. For some reason it never threw the error, the code would just continue as if all was fine. We agreed to for now use `console.error`.

aaemnnosttv commented on assets/js/modules/adsense/common/account-select.js:
```diff
@@ -0,0 +1,69 @@
+		}
+	} );
+
+	if ( 'undefined' === typeof accounts ) {
```
Let's use `accounts === undefined` rather than `typeof`

aaemnnosttv commented on assets/js/modules/adsense/common/account-select.js:
```diff
@@ -0,0 +1,69 @@
+	} );
+
+	if ( 'undefined' === typeof accounts ) {
+		return <ProgressBar small />;
```
In the future I think it could be cool to have component-specific loading components to better hint at what the pending UI is but also to maintain consistency between loading and loaded layouts.

aaemnnosttv commented on assets/js/modules/adsense/common/ad-blocker-warning.js:
```diff
@@ -0,0 +1,73 @@
+				{ [ `googlesitekit-settings-module-warning--${ context }` ]: context }
+			) }
+		>
+			<SvgIcon id="error" height="20" width="23" /> { message }
```
Should `message` not be within an element of some kind?

felixarntz replied:
This was like that originally, so I kept it unchanged. I don't feel strongly about it, but maybe we should keep it as is for now.

aaemnnosttv commented on assets/js/modules/adsense/common/ad-blocker-warning.js:
```diff
@@ -0,0 +1,73 @@
+		<div
+			className={ classnames(
+				'googlesitekit-settings-module-warning',
+				{ [ `googlesitekit-settings-module-warning--${ context }` ]: context }
```
What is `context` here?

felixarntz replied:
This is some historic thing, there are different CSS classes for the ad blocker warning rendered in different areas. The `context` itself is passed by legacy code that renders the warning in different places, see the `googlesitekit.ModuleSettingsWarning` filter.

aaemnnosttv commented on assets/js/modules/adsense/common/error-notice.js:
```diff
@@ -0,0 +1,38 @@
+
+	// Do not display if no error or if the error yields an account status,
+	// in which case it is an "expected" error.
+	if ( ! error || 'undefined' !== typeof errorToStatus( error ) ) {
```
Let's use `!== undefined` instead
```suggestion
	if ( ! error || errorToStatus( error ) !== undefined ) {
```

tofumatt replied:
See my earlier comment about `undefined` 🙂

aaemnnosttv commented on assets/js/modules/adsense/common/site-steps.js:
```diff
@@ -0,0 +1,83 @@
+	// Site status is only displayed in sites list view, so do not pass siteURL here.
+	const siteStatusURL = getAccountSiteURL( { accountID, userEmail } );
+
+	const steps = [
+		{
+			linkText: __( 'Enabled auto ads for your site', 'google-site-kit' ),
+			linkURL: enableAutoAdsURL,
+		},
+		{
+			linkText: __( 'Verified that your site is marked as Ready', 'google-site-kit' ),
+			linkURL: siteStatusURL,
+		},
+	];
```
One alternative to consider here would be each object as props to pass to the `<Link>` below (e.g. `<Link external inherit { ...props } />`. So instead of `linkURL` it would be `href`, and `linkText` would be `children`. Just an idea which would give a bit more flexibility to each link, but what you have is fine for now.

felixarntz replied:
That's fair. I think we can keep as is for now though, since the current names of these properties are currently more clear on the other hand and less coupled to using the `Link` component. So it's a trade-off I think.

tofumatt replied:
I'm not a big fan of `prop`-spreading anyway, so I prefer the current approach 🙂

aaemnnosttv commented on assets/js/modules/adsense/common/use-snippet-switch.js:
```diff
@@ -0,0 +1,98 @@
+
+	return (
+		<Fragment>
+			<div className="googlesitekit-setup-module__switch">
```
This has a "setup" class but this component is in "common" - is this intentional?

felixarntz replied:
This is an existing CSS class from before and it was used across both setup and settings with this class. I agree it's a bit confusing, but yeah this is a "common" component regardless of this class.

aaemnnosttv commented on assets/js/modules/adsense/common/use-snippet-switch.js:
```diff
@@ -0,0 +1,98 @@
+		}
+	}, [ useSnippet ] );
+
+	if ( 'undefined' === typeof useSnippet ) {
```
```suggestion
	if ( useSnippet === undefined ) {
```

tofumatt replied:
Could also be simplified as:
```suggestion
	if ( undefined === useSnippet ) {
```

aaemnnosttv commented on assets/js/modules/adsense/common/use-snippet-switch.js:
```diff
@@ -0,0 +1,98 @@
+export default function UseSnippetSwitch( props ) {
+	const {
+		label = __( 'Let Site Kit place code on your site', 'google-site-kit' ),
+		checkedMessage,
+		uncheckedMessage,
```
Why pass these as props?

felixarntz replied:
These messages are based on the context of where we display this component. While we could generally handle existing tag cases in here potentially and update the message, there is another nuance that users should see different messaging based on whether an account was already approved before setting up with Site Kit or whether it got approved at some point during setup in Site Kit. I think that is way too specific to bake into this generally reusable component, and the copy from setup also doesn't fully apply to settings.

aaemnnosttv commented on assets/js/modules/adsense/common/use-snippet-switch.test.js:
```diff
@@ -0,0 +1,84 @@
+	} );
+
+	it( 'should render nothing when useSnippet is undefined', async () => {
+		muteConsole( 'warn' );
```
Why is this warning raised? Can we prevent it from being raised rather than surpressing it?

felixarntz replied:
Good point, I wanted to avoid the `GET modules/adsense/data/settings` call pollute the logs. But we can also set `useSnippet` to `undefined` without causing this to be triggered, so I've made that change.

aaemnnosttv commented on assets/js/modules/adsense/common/user-profile.js:
```diff
@@ -0,0 +1,52 @@
+	const userEmail = useSelect( ( select ) => select( userStoreName ).getEmail() );
+	const userPicture = useSelect( ( select ) => select( userStoreName ).getPicture() );
+
+	if ( 'undefined' === typeof userEmail || 'undefined' === typeof userPicture ) {
```
```suggestion
	if ( userEmail === undefined || userPicture === undefined ) {
```

tofumatt replied:
```suggestion
	if ( undefined == userEmail || undefined === userPicture ) {
```

aaemnnosttv commented on assets/js/modules/adsense/common/user-profile.js:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	return (
+		<p className="googlesitekit-setup-module__user">
```
Is the setup class necessary? This should probably be a `googlesitekit-adsense-` class like the others.

felixarntz replied:
This was also an existing class with all styling applied, I didn't want to touch CSS as part of this refactoring. This component here in particular is also not even purely AdSense-related. It's currently here, but potentially this could even be an overarching common component since this is just a user bubble. I just put it here for now.

tofumatt commented on assets/js/modules/adsense/common/ad-blocker-warning.test.js:
```diff
@@ -0,0 +1,48 @@
+	it( 'should render the warning when an AdBlocker is active', async () => {
+		const { container } = render( <AdBlockerWarning />, { setupRegistry: setupAdBlockerRegistry } );
+
+		expect( container.querySelector( '.googlesitekit-settings-module-warning' ) ).not.toEqual( null );
```
Testing to only not equaling `null` is a bit dangerous, since `undefined` or `''` would pass this test. It's not exactly doing what the test says; we should use `getByText` here to find the actual warning text and ensure it's rendered.

It's best not to us `querySelector` tests in general, as they're testing for what the machine sees, rather than the user, so it's a good sign that the test could be improved. 🙂

tofumatt commented on assets/js/modules/adsense/common/ad-blocker-warning.test.js:
```diff
@@ -0,0 +1,48 @@
+	it( 'should render nothing when no AdBlocker is active', async () => {
+		const { container } = render( <AdBlockerWarning />, { setupRegistry: setupNoAdBlockerRegistry } );
+
+		expect( container.querySelector( '.googlesitekit-settings-module-warning' ) ).toEqual( null );
```
```suggestion
		expect( container.firstChild ).toEqual( null );
```

Might be more appropriate here, as we want to ensure nothing is rendered. There might be better matchers for that too, I always forget what's available.

Could be worth testing for potential warning text values and ensuring they aren't rendered too.

tofumatt commented on assets/js/modules/adsense/common/site-steps.js:
```diff
@@ -0,0 +1,83 @@
+	const siteURL = useSelect( ( select ) => select( siteStoreName ).getReferenceSiteURL() );
+	const userEmail = useSelect( ( select ) => select( userStoreName ).getEmail() );
+
+	if ( ! accountID || ! siteURL || ! userEmail ) {
```
You're checking for false-y values here, but shouldn't these be checks for `undefined`? If `accountID` is `0`, this would fail, no?

felixarntz replied:
Not a big deal, I did the false-y check here intentionally, since `accountID` of `0` is not a valid account ID anyway, it's always a non-empty string.

tofumatt replied:
Oh, right, cool. 👍

tofumatt commented on assets/js/modules/adsense/common/site-steps.js:
```diff
@@ -0,0 +1,83 @@
+
+	const steps = [
+		{
+			linkText: __( 'Enabled auto ads for your site', 'google-site-kit' ),
```
Since these are sentences, they should end with full-stops, no?
```suggestion
			linkText: __( 'Enabled auto ads for your site.', 'google-site-kit' ),
```

felixarntz replied:
They aren't really full sentences I guess. I think periods should be omitted here, this was the original copy, and also for UX purposes labels or list items like this should not have them.

tofumatt commented on assets/js/modules/adsense/common/site-steps.js:
```diff
@@ -0,0 +1,83 @@
+			linkURL: enableAutoAdsURL,
+		},
+		{
+			linkText: __( 'Verified that your site is marked as Ready', 'google-site-kit' ),
```
```suggestion
			linkText: __( 'Verified that your site is marked as ready.', 'google-site-kit' ),
```

The capital "r" here is a bit strange. I'd recommend removing it, or quoting it so it's clear it's a "Status". 😄

felixarntz replied:
I'll go with the quotes.

tofumatt commented on assets/js/modules/adsense/common/use-snippet-switch.js:
```diff
@@ -0,0 +1,98 @@
+
+export default function UseSnippetSwitch( props ) {
+	const {
+		label = __( 'Let Site Kit place code on your site', 'google-site-kit' ),
```
Should we clarify the kind of code being placed here? I know in this _context_ it's AdSense, but that context might be lost on a screenreader tabbing through interactive elements.

```suggestion
		label = __( 'Let Site Kit place AdSense code on your site.', 'google-site-kit' ),
```

tofumatt commented on assets/js/modules/adsense/common/use-snippet-switch.js:
```diff
@@ -0,0 +1,98 @@
+					checked={ useSnippet }
+					disabled={ isDoingSaveUseSnippet }
+					hideLabel={ false }
+				/> <span className="googlesitekit-recommended">{ __( 'RECOMMENDED', 'google-site-kit' ) }</span>
```
Why is this text uppercase like this? If 100% desired we should do that with CSS, because this will confuse some screen readers (they might read this as an acronym rather than a word). Plus: it's not very friendly to dyslexic users and others who find all-caps text harder-to-scan.

Still: at the very least it's not semantically all-caps, so we should do this with CSS.

aaemnnosttv replied:
I had a similar thought here but I recently learned that some screen readers may still see it as all caps if transformed with CSS. I agree that doing with CSS is probably the better way to go here though 👍

tofumatt replied:
Yeah, I think doing `small-caps` or something else is better still, but ultimately any changes in case has the potential to mess with assistive technology. All-caps is generally not very accessible 😬

tofumatt commented on assets/js/modules/adsense/common/use-snippet-switch.test.js:
```diff
@@ -0,0 +1,84 @@
+
+describe( 'UseSnippetSwitch', () => {
+	afterEach( () => apiFetchMock.mockClear() );
+	afterAll( () => jest.restoreAllMocks() );
```
Why restore all mocks rather than the particular `apiFetchMock` made here? This seems like it could affect other tests is all.

tofumatt commented on assets/js/modules/adsense/common/use-snippet-switch.test.js:
```diff
@@ -0,0 +1,84 @@
+		muteConsole( 'warn' );
+		const { queryAllByRole } = render( <UseSnippetSwitch /> );
+
+		expect( queryAllByRole( 'switch', { hidden: true } ) ).toHaveLength( 0 );
```
You could test for `!container.firstChild` as well to ensure nothing is rendered.

tofumatt commented on assets/js/modules/adsense/common/use-snippet-switch.js:
```diff
@@ -0,0 +1,98 @@
+	const onChange = useCallback( async () => {
+		setUseSnippet( ! useSnippet );
+		trackEvent( 'adsense_setup', useSnippet ? 'adsense_tag_enabled' : 'adsense_tag_disabled' );
+		if ( saveOnChange ) {
```
This is confusing—why wouldn't we want this setting to _always_  update? It's a better UX than having to click a save button somewhere. It might be nice to document why this is here in the code at least.

felixarntz replied:
This is only here for legacy reasons. We don't update settings immediately anywhere, and while this is generally a good practice and worth a consideration, we shouldn't add it at this point. The only reason this prop exists is that there is one place specifically in the AdSense setup where this value needs to update right away. Elsewhere (e.g. settings) it shouldn't.

tofumatt commented on assets/js/modules/adsense/common/use-snippet-switch.js:
```diff
@@ -0,0 +1,98 @@
+};
+
+UseSnippetSwitch.defaultProps = {
+	saveOnChange: false,
```
I feel like this shouldn't be the default; I'd want the component to have the "best" UX for a user by default, and that's the "I don't have to click a 'save' button" UX 😄

Maybe this should be an override to NOT save on change, which I assume is for set up when there's nothing saved in the database yet?

felixarntz replied:
Since our default practice is to _not_ update settings on-the-fly, I think having that as the default makes sense.

tofumatt commented on assets/js/modules/adsense/common/user-profile.js:
```diff
@@ -0,0 +1,52 @@
+			<img
+				className="googlesitekit-setup-module__user-image"
+				src={ userPicture }
+				alt={ __( 'User Avatar', 'google-site-kit' ) }
```
This is not a great alt attribute; it doesn't actually describe the "content" of the image, and so doesn't provide a useful/equivalent bit of info for a non-sighted user. If we can't _describe_ the contents of the image, then we shouldn't give it alt text at all.
```suggestion
				alt=""
```

tofumatt commented on assets/js/modules/adsense/dashboard/dashboard-zero-data.js:
```diff
@@ -0,0 +1,54 @@
+					</h3>
+
+					<p>
+						{ __( 'You don’t have any ad impressions yet. Meanwhile, make sure you’ve completed these steps in AdSense.', 'google-site-kit' ) }
```
I think this copy can be simplified a bit.
```suggestion
						{ __( 'You don’t have any ad impressions yet. Make sure you’ve completed these steps in AdSense:', 'google-site-kit' ) }
```

tofumatt commented on assets/js/modules/adsense/datastore/adblocker.test.js:
```diff
@@ -0,0 +1,111 @@
+			} );
+
+			it( 'receives and sets isAdBlockerActive', async () => {
+				await dispatch.receiveIsAdBlockerActive( true );
```
Does this test need to be async? I don't think it does (not a big deal, but just saying).

tofumatt commented on assets/js/modules/adsense/datastore/adblocker.test.js:
```diff
@@ -0,0 +1,111 @@
+				if ( ! global.googlesitekit ) {
+					global.googlesitekit = {};
+				}
+				global.googlesitekit.canAdsRun = false;
```
```suggestion
				// Setting `canAdsRun` to `false` means an ad-blocker is enabled;
				// if we could run ads, then we don't have an ad-blocker on.
				global.googlesitekit.canAdsRun = false;
```

tofumatt commented on assets/js/modules/adsense/datastore/adblocker.test.js:
```diff
@@ -0,0 +1,111 @@
+				}
+				global.googlesitekit.canAdsRun = false;
+
+				// Set value to false, contrary to the global above which would result in true.
```
Hah, that's kind of confusing! See my suggestion above to clarify this a bit more; it made me do a triple-take 😅

tofumatt commented on assets/js/modules/adsense/datastore/index.js:
```diff
@@ -105,6 +140,8 @@ export const selectors = Data.collectSelectors(
 	report.selectors,
 	tags.selectors,
 	urlchannels.selectors,
+	settings.selectors,
```
This file can be simplified with `Data.combineStores()`, since it's available in `develop` now.

tofumatt commented on assets/js/modules/adsense/datastore/tags.js:
```diff
@@ -293,27 +293,43 @@ export const selectors = {
 	},
 
+	/**
+	 * Checks whther the user has access to the existing AdSense tag.
```
```suggestion
	 * Checks whether the user has access to the existing AdSense tag.
```

felixarntz replied:
Fixed both.

tofumatt commented on assets/js/modules/adsense/datastore/tags.js:
```diff
@@ -293,27 +293,43 @@ export const selectors = {
+	 * @since n.e.x.t
+	 *
+	 * @return {(boolean|null|undefined)} true or false if tag permission is available,
+	 *                    null if no existing tag,
```
The indentation on this seems a bit odd; it isn't lined up with anything.

tofumatt commented on assets/js/modules/adsense/datastore/tags.js:
```diff
@@ -326,7 +342,7 @@ export const selectors = {
 	 * @param {Object} state    Data store's state.
 	 * @param {string} clientID The AdSense Client ID to check permissions for.
-	 * @return {?boolean} True if the user has access, false if not; `undefined` if not loaded.
+	 * @return {boolean|undefined} True if the user has access, false if not; `undefined` if not loaded.
```
Multiple return types should be wrapped in parens for JSDoc:
```suggestion
	 * @return {(boolean|undefined)} True if the user has access, false if not; `undefined` if not loaded.
```

tofumatt commented on assets/js/modules/adsense/settings/settings-edit.js:
```diff
@@ -0,0 +1,104 @@
+	const isDoingSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).isDoingSubmitChanges() );
+	const haveSettingsChanged = useSelect( ( select ) => select( STORE_NAME ).haveSettingsChanged() );
+
+	// Rollback any temporary selections to saved values on dismount.
```
Is this the older rollback solution we were using—eg. didn't @aaemnnosttv update this in Analytics for an approach that involved a component higher-up in the tree? (I think it's now in `AnalyticsMain`, and would go in `SettingsMain` here.)

EDIT: Yup, see: https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics/settings/settings-main.js#L34-L42

tofumatt commented on assets/js/modules/adsense/settings/settings-main.js:
```diff
@@ -0,0 +1,35 @@
+import SettingsEdit from './settings-edit';
+import SettingsView from './settings-view';
+
+export default function SettingsMain( { isOpen, isEditing } ) {
```
I think this is the component that needs to handle the rollback, similar to: https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics/settings/settings-main.js#L34-L42

tofumatt commented on assets/js/modules/adsense/settings/settings-setup-incomplete.js:
```diff
@@ -0,0 +1,63 @@
+		<Fragment>
+			<AdBlockerWarning />
+
+			{ isPendingStatus && __( 'Site Kit has placed the code on your site, ', 'google-site-kit' ) }
```
```suggestion
			{ isPendingStatus && __( 'Site Kit has placed AdSense code on your site, ', 'google-site-kit' ) }
```

tofumatt commented on assets/js/modules/adsense/setup/setup-site-add.js:
```diff
@@ -0,0 +1,75 @@
+			<ErrorNotice />
+
+			<p>
+				{ __( 'We’ve detected that you haven’t added this site to your AdSense account yet.', 'google-site-kit' ) }
```
I don't know if the "We've detected" text really adds anything to this copy. Why don't we clarify it a bit?
```suggestion
				{ __( 'You haven’t added this site to your AdSense account yet.', 'google-site-kit' ) }
```

felixarntz replied:
This copy was defined by the AdSense team, we shouldn't change it at this point.

tofumatt commented on stories/settings.stories.js:
```diff
@@ -47,12 +51,11 @@ storiesOf( 'Settings', module )
 	} )
 	.add( 'Connected Services', () => {
 		global.googlesitekit = settingsData;
-		global.googlesitekit.setupComplete = true;
```
Why were these removed?

felixarntz replied:
Oh, good catch, these changes are leftovers related to something I temporarily changed elsewhere but then reverted. Will change this back.

adamsilverstein commented on assets/js/modules/adsense/common/account-select.js:
```diff
@@ -0,0 +1,69 @@
+		}
+	} );
+
+	if ( 'undefined' === typeof accounts ) {
```
@tofumatt & @aaemnnosttv - Thanks for pointing this out, I've also been using the older style out of habit.

tofumatt replied:
👍  It's a good idea in the pre-ES5 days when `undefined` was writable, but in modern browsers it's not something we need to worry about. I think I only got out of the habit myself recently. 😅

adamsilverstein commented:
I gave this a test in my local using my personal site where I have set up AdSense in the past to see how this affects the front end setup flow.

#### Account Detection

Account detection appeared to work smoothly with one slight issue: when I activated AdSense I get the message: "Looks like you’re already using AdSense" (right!) and the toggle to "place code on your site to get your site approved". My site is already approved though! I checked against develop and found we display the **same string** there so this is not a regression. Possibly this is because I have had AdSense disabled on my live site so I have "zero data". I have re-enabled and will test again once I have data.

![image](https://user-images.githubusercontent.com/2676022/81831786-0502e780-94fb-11ea-97f1-bdc574cca39b.png)

One other visual change from develop is that the status message about what the plugin was doing "Checking account", "Checking domain" etc (above the loading bar) are gone. This is probably a good change and intentional, mostly noting it here as a record of the change.

#### Saving settings

When I click the "Continue" button I get an error updating settings:

![image](https://user-images.githubusercontent.com/2676022/81832121-6034da00-94fb-11ea-9df8-ff3d17b83e4a.png)

Checking the debugger, I see two request to the settings endpoint, one fails and one succeeds:

![image](https://user-images.githubusercontent.com/2676022/81832636-fbc64a80-94fb-11ea-9ef7-2ded4b4f238e.png)

aaemnnosttv commented on assets/js/modules/adsense/settings/settings-form.js:
```diff
@@ -0,0 +1,68 @@
+	const clientID = useSelect( ( select ) => select( STORE_NAME ).getClientID() );
+	const existingTag = useSelect( ( select ) => select( STORE_NAME ).getExistingTag() );
+
+	let checkedMessage, uncheckedMessage;
+	if ( existingTag && existingTag === clientID ) {
+		// Existing tag with permission.
+		checkedMessage = __( 'You’ve already got an AdSense code on your site for this account, we recommend you use Site Kit to place code to get the most out of AdSense.', 'google-site-kit' );
+		uncheckedMessage = checkedMessage;
+	} else if ( existingTag ) {
+		// Existing tag without permission.
+		checkedMessage = sprintf(
+			/* translators: %s: account ID */
+			__( 'Site Kit detected AdSense code for a different account %s on your site. For a better ads experience, you should remove AdSense code that’s not linked to this AdSense account.', 'google-site-kit' ),
+			parseAccountID( existingTag )
+		);
+		uncheckedMessage = __( 'By not placing the code, AdSense account will not show ads on your website unless you’ve already got some AdSense code.', 'google-site-kit' );
+	} else {
+		// No existing tag.
+		uncheckedMessage = __( 'By not placing the code, AdSense account will not show ads on your website unless you’ve already got some AdSense code.', 'google-site-kit' );
+	}
```
Why not move this logic into `UseSnippetSwitch` since it will have access to all the same data?

felixarntz replied:
The behavior is slightly different actually in different places, therefore this should be controlled by the component rendering the switch. The conditions that these messages display in setup/settings are a bit different. While the messages are the same currently, this could easily change as well, I think coupling this into the `UseSnippetSwitch` does not make sense here.

aaemnnosttv replied:
Ah okay. I didn't realize it was different 👍 

Would it make sense to have the default messages in the component and only override it with the prop when needed?

felixarntz replied:
Hmm not sure about that. If anything, I think maybe the regular message for when there's no existing tag applies to be a default. But I think moving that out while leaving the others here would actually make the code more confusing.

tofumatt commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+	// Update current account ID setting on-the-fly.
+	useEffect( () => {
+		// Don't do anything if setting has not loaded yet or if account ID cannot be determined.
+		if ( 'undefined' === typeof previousAccountID || 'undefined' === typeof accountID ) {
```
```suggestion
		if ( undefined === previousAccountID || undefined === accountID ) {
```

tofumatt commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+	// Update current client ID setting on-the-fly.
+	useEffect( () => {
+		// Don't do anything if setting has not loaded yet or if client ID cannot be determined.
+		if ( 'undefined' === typeof previousClientID || 'undefined' === typeof clientID ) {
```
```suggestion
		if ( undefined === previousClientID || undefined === clientID ) {
```

tofumatt commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+		if ( 'undefined' === typeof previousClientID || 'undefined' === typeof clientID ) {
+			return;
+		}
+		setClientID( clientID );
```
This effect is called every time `clientID` is updated, but then it updates `clientID`. That shouldn't cause infinite re-renders as `clientID` won't show as updated by the store so won't run this _again_, but it seems like this is run one-too-many-times already. Why `setClientID` after it's changed?

felixarntz replied:
The `clientID` value that is set here does not come from the setting, but from the client detection logic (`determineClientID`). As soon as that detection logic finds the client ID, it needs to be stored in the setting. The same applies to account ID, account status, and site status, those all have settings, but they're actually automatically determined by logic.

tofumatt commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+	// Update account status setting on-the-fly.
+	useEffect( () => {
+		// Don't do anything if account status cannot be determined (because of arguments not loaded yet).
+		if ( 'undefined' === typeof accountStatus ) {
```
```suggestion
		if ( undefined === accountStatus ) {
```

tofumatt commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+
+	// Determine client.
+	const clients = useSelect( ( select ) => select( STORE_NAME ).getClients( accountID ) );
+	const clientID = determineClientID( {
```
Since we set this as the `clientID` right away… should the state handle this initialisation after `getClients` is resolved? Rather than having a component do it, which is less reusable and strikes me as a more fragile.

felixarntz replied:
There are very specific logic flows necessary here for this to work as expected, together with the account status detection logic etc. Having it all in this one component here makes most sense I think. It won't be needed anywhere else.

tofumatt commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+	};
+
+	let viewComponent;
+	if ( 'undefined' === typeof accountStatus || ( isDoingSubmitChanges && ! isSubmittingInBackground ) || isNavigating ) {
```
```suggestion
	if ( undefined === accountStatus || ( isDoingSubmitChanges && ! isSubmittingInBackground ) || isNavigating ) {
```

tofumatt commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+					accountStatus
+				) } />;
+		}
+	} else if ( 'undefined' === typeof siteStatus ) {
```
```suggestion
	} else if ( undefined === siteStatus ) {
```

tofumatt commented on assets/js/modules/adsense/util/url.js:
```diff
@@ -0,0 +1,141 @@
+ * @param {Object}             options           Options for generating the URL.
+ * @param {(string|undefined)} options.accountID The AdSense account ID.
+ * @param {(string|undefined)} options.userEmail The Google account email address. Relevant
+ *                                    for users with multiple Google accounts.
```
This indent looks off, but it's minor.

felixarntz replied:
BTW all these indent issues are from when I changed all our `?string` to `(string|undefined)` and forgot to update the indentation. 😆

tofumatt replied:
We just can't win when it comes to our docblocks, eh? 😆

tofumatt commented on assets/js/modules/adsense/util/url.js:
```diff
@@ -0,0 +1,141 @@
+ * @param {(string|undefined)} options.accountID The AdSense account ID.
+ * @param {(string|undefined)} options.siteURL   The site URL to link to in AdSense.
+ * @param {(string|undefined)} options.userEmail The Google account email address. Relevant
+ *                                    for users with multiple Google accounts.
```
Again, just an odd indent here.

tofumatt commented on assets/js/modules/adsense/util/url.js:
```diff
@@ -0,0 +1,141 @@
+const getAccountBaseURL = ( { accountID, userEmail, path = '/home' } = {} ) => {
+	let baseURL = 'https://www.google.com/adsense/new';
+	if ( userEmail ) { // Allows to indicate which Google account to use.
+		baseURL += `/u/${ userEmail }`;
```
Should we be urlencoding this email? Or does that happen elsewhere?

felixarntz replied:
Is that necessary here for an email? Would it ever have special characters?

aaemnnosttv replied:
In my testing it doesn't seem to be necessary when used here or with `authuser` param.

tofumatt replied:
I mean, validating emails is a can of worms we don't need to get too into 😆

But my understanding is you could have an email with non-ASCII text: https://en.wikipedia.org/wiki/Email_address#Internationalization

So an email like `шел+testлы@foo.com` would result in the URL `https://www.google.com/adsense/new/u/%D1%88%D0%B5%D0%BB+test%D0%BB%D1%8B@foo.com` instead of `https://www.google.com/adsense/new/u/шел+testлы@foo.com`.

Maybe not a big deal though? 🤷

felixarntz replied:
Let's leave as is for now, we can re-evaluate if it results in problems.

tofumatt commented on stories/module-adsense-setup.stories.js:
```diff
@@ -0,0 +1,389 @@
+/**
```
AMAZING to see all these here! 👍

tofumatt commented on stories/module-analytics-settings.stories.js:
```diff
@@ -47,6 +47,12 @@ function filterAnalyticsSettings() {
 	);
 }
 
+const completeModuleData = {
```
Why was these needed? If every connected module will need it, we should probably make it a global change.

felixarntz replied:
This is only here to decouple at least a little bit from legacy behavior. All our Storybook components currently heavily rely on legacy `global.googlesitekit` data which is terrible to maintain and regenerate (see all the files in `.storybook/data`). This here is used for rely on that data for the most part, but then enforce certain module flags to ensure it is considered active and set up. In the (probably very) long term, we will not have the `global.googlesitekit` legacy data anymore, and this would all be scaffolded individually or use a fixture, a mock creator or something like that.

felixarntz commented on assets/js/modules/adsense/common/account-select.test.js:
```diff
@@ -0,0 +1,85 @@
+	registry.dispatch( STORE_NAME ).receiveAccounts( fixtures.accountsMultiple );
+};
+
+const setupLoadingRegistry = ( registry ) => {
```
It will actually result in a registry that is in a loading state when it comes to accounts, hence the name.

felixarntz commented on stories/settings.stories.js:
```diff
@@ -22,6 +22,10 @@ import { __ } from '@wordpress/i18n';
 	.add( 'Settings Tabs', () => {
+		global.googlesitekit = settingsData;
+		global.googlesitekit.modules.adsense.active = false;
+		global.googlesitekit.modules.adsense.setupComplete = false;
```
This code shouldn't be here anyway - note though that this `setupComplete` does not reference the AdSense-specific setting but the general flag every module has.

aaemnnosttv replied:
Oh right 😄

felixarntz commented on assets/js/modules/adsense/datastore/__fixtures__/accounts-multiple.json:
```diff
@@ -0,0 +1,18 @@
+[
```
Good idea, let's think about it in a separate issue.

felixarntz commented on assets/js/modules/adsense/datastore/adblocker.js:
```diff
@@ -0,0 +1,107 @@
+		// Global is set by ads.js entry point script.
+		const canAdsRun = global.googlesitekit && global.googlesitekit.canAdsRun;
+
+		yield actions.receiveIsAdBlockerActive( ! canAdsRun );
```
I think this is more of a preference, generally I'd say early returns like here are easier to read than what we have in many other places.

felixarntz commented on assets/js/modules/adsense/datastore/adblocker.test.js:
```diff
@@ -0,0 +1,111 @@
+				}
+				global.googlesitekit.canAdsRun = true;
+
+				expect( global.googlesitekit.canAdsRun ).not.toEqual( undefined );
```
Oops 🤷‍♂️

felixarntz commented on assets/js/modules/adsense/datastore/clients.js:
```diff
@@ -161,9 +161,24 @@ export const reducer = ( state, { type, payload } ) => {
+			} = state.savedSettings || {};
 			return {
 				...state,
 				clients: {},
```
Good idea, I've added that to the Analytics store `RESET_ACCOUNTS` too.

felixarntz commented on assets/js/modules/adsense/datastore/index.js:
```diff
@@ -62,6 +69,17 @@ export const actions = Data.addInitializeAction( Data.collectActions(
+				type: 'RECEIVE_ERROR',
+				payload: { error },
+			};
+		},
```
True, but this is good to have for testing, we actually need to use it in AdSense in some places because certain account states rely on expected errors. I'd argue we should potentially add this to other stores too (not as part of this PR).

felixarntz commented on assets/js/modules/adsense/datastore/settings.test.js:
```diff
@@ -0,0 +1,351 @@
+import API from 'googlesitekit-api';
+import { STORE_NAME } from './constants';
+import {
+	ACCOUNT_STATUS_APPROVED,
+	SITE_STATUS_ADDED,
```
The `constants` file is for constants for the store, but these constants are more general and not tied to the store at all. That's why I defined them in the status utility file that they most apply to.

felixarntz commented on assets/js/modules/adsense/datastore/settings.test.js:
```diff
@@ -0,0 +1,351 @@
+
+	beforeEach( () => {
+		registry = createTestRegistry();
+		dispatch = registry.dispatch( STORE_NAME );
+		select = registry.select( STORE_NAME );
```
Hmm I'm not sure I agree. Since `select` and `dispatch` are such common language and function names one should be familiar with when using datastores, I think having the variables called exactly that indicates more clearly what they are than using any custom names, even if they are maybe more semantically correct.

aaemnnosttv replied:
My point is more about consistency, especially when the definition and usage are separated in the file. `select` and `dispatch` are common language in the codebase, but always as functions. I can see how it might be confusing for some to see this used as functions/methods in one place and variables in another. We should avoid having `select` and `dispatch` scoped to the current store (either as variables or functions) as it is is inconsistent with other usage in the codebase.

Perhaps the best solution here might be to use the full `registry` select/dispatch in each test and pull out the selectors or actions we need. This yields the most clarity in definition (where they come from and what they are) and still concise in use. I would prefer this over other options as this approach is used in the source code as well (e.g. `canSubmitChanges`).

```js
const { isAdBlockerActive } = registry.select( STORE_NAME );
const { receiveIsAdBlockerActive } = registry.dispatch( STORE_NAME );
```

I'm not suggesting we do this everywhere, as it would actually be more code to write in cases with fewer uses, but it could help clean things up in tests where the repetition gets overly verbose.

felixarntz commented on assets/js/modules/adsense/datastore/settings.test.js:
```diff
@@ -0,0 +1,351 @@
+
+		describe( 'fetchSaveUseSnippet', () => {
+			it( 'requires the useSnippet param', () => {
+				const consoleErrorSpy = jest.spyOn( global.console, 'error' );
+
+				dispatch.fetchSaveUseSnippet();
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'useSnippet is required.' );
+
+				consoleErrorSpy.mockClear();
```
See my other comment regarding that. It's not too big of a deal, I think we should explore on how to deal with this as part of #1288.

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+		// Require account ID to be either empty (if impossible to determine)
+		// or valid.
+		const accountID = getAccountID();
+		if ( '' !== accountID && ! isValidAccountID( accountID ) ) {
+			return false;
+		}
+		// Require client ID to be either empty (if impossible to determine)
+		// or valid.
+		const clientID = getClientID();
+		if ( '' !== clientID && ! isValidClientID( clientID ) ) {
```
It is specific to here at this point, this is the only place where we check validity for the values to be saved. This is a bit different than GA because these values are typically not selected. The only reason they need to allow empty is for specific account statuses where having an account ID or client ID will be impossible (but the account status still needs to be saved).

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+		return true;
+	} ),
+
+	isDoingSubmitChanges( state ) {
+		return !! state.isDoingSubmitChanges;
+	},
+
+	isDoingSaveUseSnippet( state ) {
```
Updated in GA store `settings` too.

felixarntz commented on stories/notifications.stories.js:
```diff
@@ -13,22 +13,26 @@ import { __ } from '@wordpress/i18n';
+		global.googlesitekit.canAdsRun = true;
+		global.googlesitekit.modules.adsense.active = false;
+		global.googlesitekit.modules.adsense.setupComplete = false;
+
```
Accidentally here, I've reverted.

felixarntz commented on assets/js/modules/adsense/util/status.js:
```diff
@@ -17,11 +17,11 @@
  */
 
 export const ACCOUNT_STATUS_NONE = 'none';
+export const ACCOUNT_STATUS_MULTIPLE = 'multiple';
 export const ACCOUNT_STATUS_DISAPPROVED = 'disapproved';
 export const ACCOUNT_STATUS_GRAYLISTED = 'graylisted';
```
See my other comment, `constants.js` is specific to the datastore while these are not.

felixarntz commented on assets/js/modules/adsense/setup/setup-site-add.js:
```diff
@@ -0,0 +1,75 @@
+import Button from '../../../components/button';
+import { getAccountSiteURL } from '../util/url';
+import { STORE_NAME } from '../datastore/constants';
+import { STORE_NAME as siteStoreName } from '../../../googlesitekit/datastore/site/constants';
+import { STORE_NAME as userStoreName } from '../../../googlesitekit/datastore/user/constants';
```
I think we're using mixed variants throughout the codebase, let's revisit this separately. I'm not sure about which naming convention is better, maybe there's another one (e.g. `CORE_SITE_STORE` or `STORE_CORE_SITE`?).

aaemnnosttv replied:
I would opt for `STORE_CORE_SITE` from those two. Definitely think we should preserve the upper snake case though.

felixarntz commented on assets/js/modules/adsense/setup/setup-main.js:
```diff
@@ -0,0 +1,293 @@
+		}
+		// Set internal state for submitting in background to avoid sudden
+		// rendering of a progress bar.
+		( async () => {
+			setIsSubmittingInBackground( true );
+			await submitChanges();
+			setIsSubmittingInBackground( false );
+		} )();
```
I tried that initially, but `useEffect` apparently does not support async callbacks.

aaemnnosttv replied:
Ah, I thought we were using it elsewhere in the code but it was `useCallback` 😄 

That makes sense; this is a bit cleaner than the promise-based alternative.

felixarntz commented on assets/js/modules/adsense/settings/settings-setup-incomplete.js:
```diff
@@ -0,0 +1,63 @@
+
+export default function SettingsSetupIncomplete() {
+	const accountStatus = useSelect( ( select ) => select( STORE_NAME ).getAccountStatus() );
+	const isPendingStatus = accountStatus === ACCOUNT_STATUS_GRAYLISTED || accountStatus === ACCOUNT_STATUS_PENDING;
```
I've added `isPendingAccountStatus` in `status.js` (used in two places).

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+};
+
+export const actions = {
+	*fetchSaveUseSnippet( useSnippet ) {
```
This is per our naming conventions for fetch calls that we've defined. All `fetch*` actions should be use the original action or selector as its name.

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+		let response, error;
+
+		if ( 'undefined' === typeof useSnippet ) {
+			error = {
```
Hmm not sure what the benefit here is. Since we use plain objects overall, I think it's okay to do that here as well.

aaemnnosttv replied:
I haven't seen that but it's not a bad idea since `error` is usually set from the catch block, it will usually be an instance of `Error` rather than a plain object.

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+	 * Saves the current value of the 'useSnippet' setting.
+	 *
+	 * While the saveSettings action should typically be used for this, there
+	 * is a use-case where the 'useSnippet' setting (and nothing else) needs to
```
This is required for the `UseSnippetSwitch`'s `saveOnChange` controlled logic to work (it should only save the value of `useSnippet` and not any other potentially pending changes). I've annotated it to be private, good point 👍

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+		return true;
+	} ),
+	[ COMPLETE_SITE_SETUP ]: createRegistryControl( ( registry ) => async () => {
+		await registry.dispatch( STORE_NAME ).setSiteSetupComplete( true );
```
Both `accountSetupComplete` and `siteSetupComplete` are settings (flags) that need to be set to indicate the user has completed these two respective parts of the setup. In case saving these "complete" settings fails for some reason, the values should be set back to `false`, to avoid having our client functionality behave as if setup was complete when it actually isn't.

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+			return {
+				...state,
+				// Update saved settings.
+				savedSettings: {
```
We can maybe revisit this separately, the problem is that we don't have a way to set "saved settings", this is fully internal right now, and should generally remain this way, per our previous discussion at some point. I agree this code is a bit risky if we changed the way `createSettingsStore` would work, but I think this is okay for this special case.

tofumatt replied:
That's fine, but in that case let's just make sure that there are lots of tests around this code using public selectors. I'm fine with us keeping this as-is if we only end up doing it in a few places, but we should get notified if changing how the settings store's internals breaks it 😄

felixarntz replied:
There's one that should cover this already, but I'll add an extra test that ensures the data is consistent.

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+		// This action is mainly handled via createSettingsStore, but here we
+		// need it to have the side effect of storing the original account
+		// status.
+		case 'RECEIVE_SETTINGS': {
```
These values are not exported, and they're actually generated inside a function. I think this here is a special case, so let's keep as is for now.

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+		// Require account ID to be either empty (if impossible to determine)
+		// or valid.
+		const accountID = getAccountID();
+		if ( '' !== accountID && ! isValidAccountID( accountID ) ) {
```
This is a very specific check here, an empty string is not a valid account ID, so that shouldn't be baked into the function. However, we need to allow for an empty string to be saved, since there are cases where no account ID (same for client ID) can be determined.

felixarntz commented on assets/js/modules/adsense/datastore/settings.js:
```diff
@@ -0,0 +1,444 @@
+		// Require client ID to be either empty (if impossible to determine)
+		// or valid.
+		const clientID = getClientID();
+		if ( '' !== clientID && ! isValidClientID( clientID ) ) {
```
See above.

felixarntz commented on assets/js/modules/adsense/datastore/settings.test.js:
```diff
@@ -0,0 +1,351 @@
+				dispatch.setSettings( validSettings );
+				expect( select.canSubmitChanges() ).toBe( true );
+
+				dispatch.setAccountID( '0' );
```
Good point, I've also added `''` specifically, since that _should_ actually allow to submit changes, despite not being a valid ID.

felixarntz commented on assets/js/modules/adsense/settings/settings-view.js:
```diff
@@ -0,0 +1,129 @@
+
+			<div className="googlesitekit-settings-module__meta-items">
+				<div className="googlesitekit-settings-module__meta-item">
+					<p className="googlesitekit-settings-module__meta-item-type">
```
Let's open a separate issue for this. This has been bugging me as well for a long time, but it goes throughout all modules, so shouldn't be done here. @tofumatt Care to open one? :)

tofumatt replied:
Filed #1557, though I didn't get the chance to write-up the IB yet in the interest of time 😄

felixarntz commented on assets/js/modules/adsense/setup/setup-account-approved.js:
```diff
@@ -0,0 +1,131 @@
+		return null;
+	}
+
+	const isApprovedFromVeryBeginning = '' === originalAccountStatus || ACCOUNT_STATUS_APPROVED === originalAccountStatus;
```
I've added a comment explaining.

felixarntz commented on assets/js/modules/adsense/setup/setup-account-approved.js:
```diff
@@ -0,0 +1,131 @@
+				}
+			</h3>
+
+			<ErrorNotice />
```
Yep, it handles that internally.

felixarntz commented on assets/js/modules/adsense/setup/setup-account-disapproved.js:
```diff
@@ -0,0 +1,68 @@
+	const accountID = useSelect( ( select ) => select( STORE_NAME ).getAccountID() );
+	const userEmail = useSelect( ( select ) => select( userStoreName ).getEmail() );
+
+	if ( ! userEmail ) {
```
Good catch! `accountID` may be an empty string here though (which is okay here), so I'll use an `undefined` check.

felixarntz commented:
@aaemnnosttv @tofumatt I think I've addressed all your feedback (or left comments why I may not have), thanks for all that. I've also implemented a few other things that were previously missing:
* The background submit logic in `SetupMain` should only save settings in the background if one of account ID, client ID, account status, or site status have changed. So far this was not working correctly, it would also update e.g. when you toggle the snippet switch, plus there were some sync issues. What I did to solve is set a flag whenever one of these settings is updated to indicate we're awaiting a submission. Once all requirements are fulfilled (via `canSubmitChanges`), the submission then actually happens. See https://github.com/google/site-kit-wp/pull/1410/commits/cadfedc6bf3c840ae2627050074443f19552ced1
* The `saveOnChange` functionality of the snippet switch, which should be used during setup was not actually present as a prop. It would previously still have saved the setting but only due to the above bug, and it could have had unintended side effects. Now with the prop passed, it's ensured that only the `useSnippet` setting is saved. See https://github.com/google/site-kit-wp/pull/1410/commits/f93e6b53740eb12f29fadf9a4e0f8244523bf363
* I've added the refresh logic that had been there in the original implementation: If the user focuses elsewhere and comes back and 15 seconds or more have passed, all data should be re-fetched (the timing is same like before, let's keep that). The original logic was removed (while it was general logic, it was only used by AdSense, and we're certainly not going to use it anywhere else). See https://github.com/google/site-kit-wp/pull/1410/commits/aa09d3793ddcb124520b560e4b6e202e1c32f6bd

Would be great if you could check these individual commits specifically. The one thing missing now is fixing the Storybook hacks, I didn't get to that yet. Maybe we can discuss this tomorrow.

@adamsilverstein Thanks for the testing, yeah the message there "to get the site approved" is currently expected with the toggle, per definition with AdSense team. Good point on the lack of the loading messages (we don't do that elsewhere either); I honestly didn't think about that, but I also think it's perfectly fine to not have them. Regarding the 500 error you see from saving settings, could you re-check with the latest commit? Potentially there was a temporary fatal at some commit. If it still happens, we need to dive deeper into why the server is giving this error there.

aaemnnosttv commented:
@felixarntz I'm seeing a bit of lag after the initial display of the progress bar and the next components to show. In the video, this is just before it says "Looks like you're already using AdSense". Take a look here:

![Kapture 2020-05-14 at 19 00 32](https://user-images.githubusercontent.com/1621608/81957599-7931ae00-9615-11ea-8a56-e80862b5d23a.gif)

felixarntz commented:
@aaemnnosttv Hmm, curious what piece of information it's waiting for there. Should we modify `SetupMain` to ensure _all_ data that child components need is loaded, even if `SetupMain` may not otherwise need it itself? I think you did that for Analytics, but I wasn't sure about it since it kind of mixes up concerns. But I also don't see a better way currently.

felixarntz commented on assets/js/modules/adsense/setup/setup-site-added.js:
```diff
@@ -0,0 +1,96 @@
+			return;
+		}
+
+		// TODO: Remove temporary hack to avoid saving in Storybook.
+		if ( global.__STORYBOOK_ADDONS ) {
+			finishSetup();
+			trackEvent( 'adsense_setup', 'complete_adsense_setup' );
+			return;
+		}
```
The hack is only here to not call `completeSiteSetup` below in Storybook, which would issue an API request. But either way, this should be removed before this makes it into a release, and potentially replaced with a fetch mock.

felixarntz commented:
@aaemnnosttv The problem you raised above should be fixed via https://github.com/google/site-kit-wp/pull/1410/commits/60fb74ee3e1f32a7b30b146f3c10b58375cc8314. I've also addressed a similar issue with settings in https://github.com/google/site-kit-wp/pull/1410/commits/039271394ec82172e74e93f415a9af48ecb0d2d3.
